File: sdks/elixir/lib/autonoma/dialect.ex
defmodule Autonoma.Dialect do
  @moduledoc "Database dialect abstraction — generates dialect-specific SQL strings."

  defmodule Postgres do
    @moduledoc false

    alias Autonoma.Generated.SQLQueries

    def name, do: "postgres"
    def supports_returning, do: true
    def param(i), do: "$#{i}"
    def quote_id(name), do: ~s("#{name}")

    def tables_sql(schema), do: Autonoma.Dialect.do_replace(SQLQueries.postgres_tables(), schema)
    def columns_sql(schema), do: Autonoma.Dialect.do_replace(SQLQueries.postgres_columns(), schema)
    def primary_keys_sql(schema), do: Autonoma.Dialect.do_replace(SQLQueries.postgres_primary_keys(), schema)
    def foreign_keys_sql(schema), do: Autonoma.Dialect.do_replace(SQLQueries.postgres_foreign_keys(), schema)
    def enums_sql(_schema), do: SQLQueries.postgres_enums()
  end

  defmodule MySQL do
    @moduledoc false

    alias Autonoma.Generated.SQLQueries

    def name, do: "mysql"
    def supports_returning, do: false
    def param(_i), do: "?"
    def quote_id(name), do: "`#{name}`"

    def tables_sql(schema), do: Autonoma.Dialect.do_replace(SQLQueries.mysql_tables(), schema)
    def columns_sql(schema), do: Autonoma.Dialect.do_replace(SQLQueries.mysql_columns(), schema)
    def primary_keys_sql(schema), do: Autonoma.Dialect.do_replace(SQLQueries.mysql_primary_keys(), schema)
    def foreign_keys_sql(schema), do: Autonoma.Dialect.do_replace(SQLQueries.mysql_foreign_keys(), schema)
    def enums_sql(_schema), do: SQLQueries.mysql_enums()
  end

  @doc false
  def do_replace(template, schema) do
    String.replace(template, "{{schema}}", schema)
  end

  def get("postgres"), do: Postgres
  def get("mysql"), do: MySQL
  def get(name), do: raise("Dialect \"#{name}\" is not yet supported.")
end

File: sdks/go/autonoma/dialect.go
package autonoma

import (
	"fmt"
	"strings"
)

// Dialect abstracts SQL differences between database engines.
type Dialect struct {
	Name              string
	Param             func(index int) string
	QuoteID           func(name string) string
	SupportsReturning bool
	TablesSQL         func(schema string) string
	ColumnsSQL        func(schema string) string
	PrimaryKeysSQL    func(schema string) string
	ForeignKeysSQL    func(schema string) string
	EnumsSQL          func(schema string) string
}

func replaceSchema(template, schema string) string {
	return strings.ReplaceAll(template, "{{schema}}", schema)
}

var PostgresDialect = Dialect{
	Name:              "postgres",
	Param:             func(i int) string { return fmt.Sprintf("$%d", i) },
	QuoteID:           func(name string) string { return `"` + name + `"` },
	SupportsReturning: true,
	TablesSQL:         func(schema string) string { return replaceSchema(POSTGRES_TABLES, schema) },
	ColumnsSQL:        func(schema string) string { return replaceSchema(POSTGRES_COLUMNS, schema) },
	PrimaryKeysSQL:    func(schema string) string { return replaceSchema(POSTGRES_PRIMARY_KEYS, schema) },
	ForeignKeysSQL:    func(schema string) string { return replaceSchema(POSTGRES_FOREIGN_KEYS, schema) },
	EnumsSQL:          func(schema string) string { return replaceSchema(POSTGRES_ENUMS, schema) },
}

var MySQLDialect = Dialect{
	Name:              "mysql",
	Param:             func(_ int) string { return "?" },
	QuoteID:           func(name string) string { return "`" + name + "`" },
	SupportsReturning: false,
	TablesSQL:         func(schema string) string { return replaceSchema(MYSQL_TABLES, schema) },
	ColumnsSQL:        func(schema string) string { return replaceSchema(MYSQL_COLUMNS, schema) },
	PrimaryKeysSQL:    func(schema string) string { return replaceSchema(MYSQL_PRIMARY_KEYS, schema) },
	ForeignKeysSQL:    func(schema string) string { return replaceSchema(MYSQL_FOREIGN_KEYS, schema) },
	EnumsSQL:          func(_ string) string { return MYSQL_ENUMS },
}

// GetDialect returns the dialect for the given name. Defaults to "postgres".
func GetDialect(name string) (*Dialect, error) {
	switch name {
	case "", "postgres":
		d := PostgresDialect
		return &d, nil
	case "mysql":
		d := MySQLDialect
		return &d, nil
	default:
		return nil, fmt.Errorf("dialect %q is not yet supported. Currently only \"postgres\" and \"mysql\" are available", name)
	}
}
